### PR TITLE
ENH: Environment Analysis

### PR DESCRIPTION
**Pull request by PatrickSampaioUSP**
It ain't ready yet but we should start discussing it

task; https://app.clickup.com/3146690/v/dc/300y2-633/300y2-1203

I'm currently using a granularity of 3 hours between each batch of information for each day. Maybe this should be revised, as some tasks already explicitly ask for a greater granularity, however if we decide to have a more granular data, we may have 2 problem:
1 - As I observed on the data from the FORECAST, the intervals between the timeIndex of the file data were about 3 hours, so to raise the granularity I would have to check it out if it is possible
2 - Time to run vs changes on the code, the current solution affects minimally the Environment class, so to get more granular data we currently have to expend more time on the processing

Expected usage:

```
In [1]: from rocketpy.EnvironmentAnalysis import EnvironmentAnalysis
   ...: from datetime import date
   ...: 
   ...: Env = EnvironmentAnalysis(railLength=5.2,
   ...:         latitude=32.990254,
   ...:         longitude=-106.974998,
   ...:         elevation=1400,
   ...:         start_date=(2022, 2, 15),
   ...:         end_date=(2022, 2, 17) # Tomorrow's date in year, month, day, hour UTC format
   ...:    )

```

Wind rose snippet from animation:

Figure_1

Weibull Distribution exmple:
weibull

Some considerations: I think we should build this code in order that it could be easily removed from RocketPy, my interpretation is that we should make a new library dedicated to Environment Analysis

### CODE REVIEW

**Comment by Gui-FernandesBR** [comment]
Adding + @MateusStano , the thematic is of your interest

**Comment by Gui-FernandesBR** [comment]
Could separate Function as well...

**Comment by Gui-FernandesBR** [comment]
According to Giovani, this is still a WiP Pull Request therefore we are not starting the review yet.
@PatrickSampaioUSP please let us to know when it's time to start reviewing it, I look forward to do that

File: rocketpy/EnvironmentAnalysis.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        self.max_temperature = np.max(
+            [np.max(env.temperature.source[:, 1]) for env in self.environments]
+        )
+
+    def calculate_min_temperature(self):
+        self.min_temperature = np.min(
+            [np.min(env.temperature.source[:, 1]) for env in self.environments]
+        )
+
+    def calculate_average_max_wind_gust(self):
```
Comment (PatrickSampaioUSP) [inline rocketpy/EnvironmentAnalysis.py:90]: read field wind-gust

**Comment by Gui-FernandesBR** [comment]
Adding @FranzYuri to this thread

**Comment by giovaniceotto** [comment]
@PatrickSampaioUSP, @Gui-FernandesBR and @FranzYuri, here is a todo list:

# TODO: Check implementation and use when parsing files
- [ ] `__bilinear_interpolation`

# TODO: Implement
- [ ] `parsePressureLevelData`
- [ ] `parseSurfaceData`
- [ ] `calculate_average_max_wind_gust`
- [ ] `calculate_maximum_wind_gust`
- [ ] `calculate_wind_gust_distribution`
- [ ] `calculate_average_temperature_along_day`
- [ ] `calculate_average_wind_profile`
- [ ] `calculate_average_day_wind_rose`
- [ ] `animate_wind_gust_distribution_over_average_day`
- [ ] `animate_wind_profile_over_average_day`

# TODO: Adapt to new data format
- [ ] `calculate_average_max_temperature`
- [ ] `calculate_average_min_temperature`
- [ ] `calculate_record_max_temperature`
- [ ] `calculate_record_min_temperature`
- [ ] `animate_wind_rose`

# TODO: Notebook
- [ ] Create a notebook to showcase the class. Use files `data\weather\spaceport_america_surface_reanalysis_2015_2021.nc` and `data\weather\spaceport_america_pressure_level_reanalysis_2015_2021.nc` as an example.

# FYI
I would suggest starting by `__bilinear_interpolation`, `parsePressureLevelData` and `parseSurfaceData` in this order. Then, all the other methods to be implemented are easier and can be done in any order. I believe the method names and docstring are all you need to get started. Let me know if you have any questions.

For now, here is the data structure I chose to keep the data parsed by `parsePressureLevelData` and `parseSurfaceData`:

```
pressureLevelDataDict: {
            "date": {
                "hour": {
                    "pressure": ...,
                    "temperature": ...
                },
                "hour": {
                    "data": ...,
                    "data": ...
                }
            },
            "date": {
                "hour": {
                    "data": ...,
                    "data": ...
                },
                "hour": {
                    "data": ...,
                    "data": ...
                }
            }
        }

surfaceDataDict: {
            "date": {
                "hour": {
                    "data": ...,
                    "data": ...
                },
                "hour": {
                    "data": ...,
                    "data": ...
                }
            },
            "date": {
                "hour": {
                    "data": ...,
                    "data": ...
                },
                "hour": {
                    "data": ...,
                    "data": ...
                }
            }
        }
```

**Comment by PatrickSampaioUSP** [comment]
At the `calculate_average_temperature_along_day` there is the following comment `"""temperature progression throughout the day at some fine interval (ex: 2 hours) with 1, 2, 3, sigma contours"""` What exactly you mean by the sigma countours in this case?

**Comment by PatrickSampaioUSP** [comment]
@giovaniceotto Also, for the windrose we will use only the windSpeed and windDIrection data of the given elevation?

**Comment by giovaniceotto** [comment]
> @giovaniceotto Also, for the windrose we will use only the windSpeed and windDIrection data of the given elevation?

We should probably use surface data for the wind rose. Would that work? So no elevation data.

**Comment by giovaniceotto** [comment]
> At the `calculate_average_temperature_along_day` there is the following comment `"""temperature progression throughout the day at some fine interval (ex: 2 hours) with 1, 2, 3, sigma contours"""` What exactly you mean by the sigma countours in this case?

These are Logan's words. I would say it means you need to display the mean value (over all days), the mean +- 1 standard deviation (over all days), the mean +- 2 standard deviation (over all days) and the mean +- 3 standard deviation (over all days).

**Comment by Gui-FernandesBR** [comment]
Where to export .nc data: 
- Surface level:  https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels?tab=overview
- Pressure level: https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-pressure-levels?tab=form

Additional:
- We need a good report to present our development to Logan and Esra team. Plot.ly dash, xarray, html, .ppt, etc.

**Comment by Gui-FernandesBR** [comment]
New set of tasks:
- @giovaniceotto --> Correct the problem Finnish 
- @FranzYuri --> Finnish temperature analysis and create draft pull request; and then start thinking about timezone
- @Gui-FernandesBR --> Understand how to implement different units of measure
- @PatrickSampaioUSP --> Conclude windrose implementation applying corrections
- @MateusStano --> Open Draft Pull Request with animations implementation so Giovani can finish it

**Comment by Gui-FernandesBR** [comment]
I think we need to re-assign  time-zones task from Franz to someone else.
Otherwise we can skip this part as it is no mandatory

**Comment by giovaniceotto** [comment]
> I think we need to re-assign time-zones task from Franz to someone else. Otherwise we can skip this part as it is no mandatory

I believe it is quite important. Analyzing the data in UTC time can be quite hard, specially during animations, which will start at 00:00 UTC.

The implementation should be simple. I will take it and submit a PR tomorrow.

**Comment by Gui-FernandesBR** [comment]
What is happening with pytest here? It seems it got an error on loading tkinter library (??)

**Review by Gui-FernandesBR — APPROVED** [review]
For me this can already be approved and merged! The only thing left is the wind heading analysis, which can be worked separated.
Also, we should start creating some unit tests in a different PR

**Review by giovaniceotto — APPROVED** [review]
Good enough to let people use it!

What needs to be done in the future:
- Add unit tests
- Allow conversion of Environment Analysis to Environment
- Optimize speed

**Pull request merged**